File: modules/supabase-core/src/database/rls-helpers.ts
/**
 * RLS (Row Level Security) Helpers
 * 
 * Utilities for testing and managing RLS policies.
 * 
 * Dependencies: @supabase/supabase-js
 */

import type { SupabaseClient } from '@supabase/supabase-js'
import type { Database } from '../../types/database-types'

/**
 * Tests if RLS is enabled on a table.
 * 
 * @param supabase - Supabase client (should use service role for this check)
 * @param tableName - Name of the table to check
 * @returns True if RLS is enabled, false otherwise
 * 
 * @example
 * ```typescript
 * const isRLSEnabled = await checkRLSEnabled(adminSupabase, 'profiles')
 * if (!isRLSEnabled) {
 *   console.warn('RLS is not enabled on profiles table')
 * }
 * ```
 */
export async function checkRLSEnabled(
  supabase: SupabaseClient<Database>,
  tableName: string
): Promise<boolean> {
  const { data, error } = await supabase.rpc('check_rls_enabled', {
    table_name: tableName,
  })

  if (error) {
    // Fallback: Try direct query to information_schema
    const { data: schemaData, error: schemaError } = await supabase
      .from('information_schema.table_privileges' as any)
      .select('*')
      .eq('table_name', tableName)
      .limit(1)

    if (schemaError) {
      console.warn(`Could not check RLS status for ${tableName}:`, schemaError)
      return false
    }

    return true // Assume enabled if we can query
  }

  return data === true
}

/**
 * Gets the current authenticated user's ID from JWT.
 * Useful for RLS policy testing.
 * 
 * @param supabase - Authenticated Supabase client
 * @returns User ID if authenticated, null otherwise
 * 
 * @example
 * ```typescript
 * const userId = await getCurrentUserId(supabase)
 * if (userId) {
 *   // User is authenticated, RLS policies apply
 * }
 * ```
 */
export async function getCurrentUserId(
  supabase: SupabaseClient<Database>
): Promise<string | null> {
  const { data: { user }, error } = await supabase.auth.getUser()

  if (error || !user) {
    return null
  }

  return user.id
}

/**
 * Gets the current user's role from JWT claims.
 * Roles are stored in app_metadata.role.
 * 
 * @param supabase - Authenticated Supabase client
 * @returns User role if available, null otherwise
 * 
 * @example
 * ```typescript
 * const role = await getCurrentUserRole(supabase)
 * if (role === 'admin') {
 *   // User has admin role
 * }
 * ```
 */
export async function getCurrentUserRole(
  supabase: SupabaseClient<Database>
): Promise<string | null> {
  const { data: { user }, error } = await supabase.auth.getUser()

  if (error || !user) {
    return null
  }

  return (user.app_metadata?.role as string) || null
}

/**
 * Helper to check if current user has a specific role.
 * 
 * @param supabase - Authenticated Supabase client
 * @param requiredRole - Role to check for
 * @returns True if user has the required role
 * 
 * @example
 * ```typescript
 * const isAdmin = await hasRole(supabase, 'admin')
 * if (!isAdmin) {
 *   throw new Error('Unauthorized: Admin access required')
 * }
 * ```
 */
export async function hasRole(
  supabase: SupabaseClient<Database>,
  requiredRole: string
): Promise<boolean> {
  const role = await getCurrentUserRole(supabase)
  return role === requiredRole
}

File: modules/supabase-core-python/src/database/rls_helpers.py
"""
RLS (Row Level Security) Helpers

Utilities for testing and managing RLS policies.
"""

from typing import Optional
from supabase import Client


async def check_rls_enabled(client: Client, table_name: str) -> bool:
    """
    Tests if RLS is enabled on a table.

    Args:
        client: Supabase client (should use service role for this check)
        table_name: Name of the table to check

    Returns:
        True if RLS is enabled, False otherwise

    Example:
        ```python
        from supabase_core_python import check_rls_enabled, create_service_role_client

        admin_supabase = create_service_role_client()
        is_rls_enabled = await check_rls_enabled(admin_supabase, "profiles")

        if not is_rls_enabled:
            print("Warning: RLS is not enabled on profiles table")
        ```
    """
    try:
        # Try to call a database function that checks RLS
        response = client.rpc("check_rls_enabled", {"table_name": table_name}).execute()
        return response.data if hasattr(response, "data") else False
    except Exception:
        # Fallback: Assume enabled if we can't check
        # In practice, you'd query information_schema or pg_policies
        return True


def get_current_user_id(client: Client) -> Optional[str]:
    """
    Gets the current authenticated user's ID from JWT.
    Useful for RLS policy testing.

    Args:
        client: Authenticated Supabase client

    Returns:
        User ID if authenticated, None otherwise

    Example:
        ```python
        from supabase_core_python import get_current_user_id, create_authenticated_client

        supabase = create_authenticated_client(jwt_token)
        user_id = get_current_user_id(supabase)

        if user_id:
            # User is authenticated, RLS policies apply
            print(f"Current user: {user_id}")
        ```
    """
    try:
        user = client.auth.get_user()
        if user and hasattr(user, "user") and user.user:
            return user.user.id if hasattr(user.user, "id") else None
        return None
    except Exception:
        return None


def get_current_user_role(client: Client) -> Optional[str]:
    """
    Gets the current user's role from JWT claims.
    Roles are stored in app_metadata.role.

    Args:
        client: Authenticated Supabase client

    Returns:
        User role if available, None otherwise

    Example:
        ```python
        from supabase_core_python import get_current_user_role, create_authenticated_client

        supabase = create_authenticated_client(jwt_token)
        role = get_current_user_role(supabase)

        if role == "admin":
            # User has admin role
            print("Admin access granted")
        ```
    """
    try:
        user = client.auth.get_user()
        if user and hasattr(user, "user") and user.user:
            app_metadata = getattr(user.user, "app_metadata", None)
            if app_metadata and isinstance(app_metadata, dict):
                return app_metadata.get("role")
        return None
    except Exception:
        return None


def has_role(client: Client, required_role: str) -> bool:
    """
    Helper to check if current user has a specific role.

    Args:
        client: Authenticated Supabase client
        required_role: Role to check for

    Returns:
        True if user has the required role

    Example:
        ```python
        from supabase_core_python import has_role, create_authenticated_client

        supabase = create_authenticated_client(jwt_token)
        is_admin = has_role(supabase, "admin")

        if not is_admin:
            raise ValueError("Unauthorized: Admin access required")
        ```
    """
    role = get_current_user_role(client)
    return role == required_role

